src/core/socialprovider.js

```diff
@@ -182,16 +182,21 @@ XMPPSocialProvider.prototype.connect = function(continuation) {
     // This may indicate a broken connection to XMPP.
     // TODO: handle this.
     this.logger.error('received unhandled close event', e);
+    if (this.status === 'ONLINE') {
+      this.ping_();
+    }
   }.bind(this));
   this.client.addListener('end', function(e) {
-    this.logger.error('received end event, status: ' + this.status, e);
-    if (this.status !== 'ONLINE') {
-      // Reject login promise.
+    if (this.status !== 'ONLINE' && this.client) {
```

Comment:
Just to clarify a bit more, there are 3 cases we want to handle for 'end' events:
* 'end' events which are the result of an intentional log out, which we want to ignore (we can detect that logout was called because this.client==null)
* 'end' events that occur before login is complete, which should print an error and cause the login promise to reject
* 'end' events that occur after login succeeds, but not due to any call to logout.  In this case we should print an error and call this.logout to cleanup state.